[DOCS] Rewrite aggs overview (#64318)
elastic/elasticsearch: docs/reference/aggregations.asciidoc

[partintro]
--
An aggregation summarizes your data as metrics, statistics, or other analytics.
Aggregations help you answer questions like:

* What's the average load time for my website?
* Who are my most valuable customers based on transaction volume?
* What would be considered a large file on my network?
* How many products are in each product category?

{es} organizes aggregations into four categories:

> * <<search-aggregations-bucket,Bucket>> aggregations that

Author: With ea072c6284f3b044ff38aa9c09587b030a6b4484, I dropped the Matrix family entirely and moved the stats matrix agg under the Metrics family. That should make things simpler for users. However, I can revert if needed. Thanks!

group documents into buckets, also called bins, based on field values, ranges,
or other criteria.

> * <<search-aggregations-matrix,Matrix>> aggregations experimental:[] that output

Reviewer: We actually said a while back that we think we can remove "experimental" on matrix: https://github.com/elastic/elasticsearch/issues/53065

So you can probably drop this.

You can specify multiple aggregations in the same request:

[source,console]
----
GET /my-index-000001/_search
{
  "aggs": {
    "my-first-agg-name": {
      "terms": {
        "field": "my-field"
      }
    },
    "my-second-agg-name": {
      "avg": {
        "field": "my-other-field"
      }
    }
  }
}
----
// TEST[setup:my_index]
// TEST[s/my-field/http.request.method/]
// TEST[s/my-other-field/http.response.bytes/]

[discrete]
[[run-sub-aggs]]
=== Run sub-aggregations

Bucket aggregations support bucket or metric sub-aggregations. For example, a
terms aggregation with an <<search-aggregations-metrics-avg-aggregation,avg>>
sub-aggregation calculates an average value for each bucket of documents. There
is no level or depth limit for nesting sub-aggregations.

[source,console]
----
GET /my-index-000001/_search
{
  "aggs": {
    "my-agg-name": {
      "terms": {
        "field": "my-field"
      },
      "aggs": {
        "my-sub-agg-name": {
          "avg": {
            "field": "my-other-field"
          }
        }
      }
    }
  }
}
----
// TEST[setup:my_index]
// TEST[s/my-field/http.request.method/]
// TEST[s/my-other-field/http.response.bytes/]

> 

Reviewer: Collapsed sections with the results could be nice here, just so folks can see them.

Author: I selectively added some more responses. However, I'm trying to avoid collapsible sections as I've gotten feedback from others that they're easy to overlook and sometimes annoying to click through.

values:

[source,console]
----
GET /my-index-000001/_search
{
  "aggs": {
    "my-agg-name": {
      "histogram": {
        "field": "my-field",
        "interval": 1000,
        "script": "_value / 1000"
      }
    }
  }
}
----
// TEST[setup:my_index]
// TEST[s/my-field/http.response.bytes/]

Some aggregations only work on specific data types. Use the `value_type`
parameter to specify a data type for a script-generated value or an unmapped
field. `value_type` accepts the following values:

* `boolean`
* `date`
* `double`, used for all floating-point numbers
* `long`, used for all integers
* `ip`
* `string`

[source,console]
----
GET /my-index-000001/_search
{
  "aggs": {
    "my-agg-name": {
      "histogram": {
        "field": "my-field",
        "interval": 1000,
        "script": "_value / 1000",
        "value_type": "long"
      }
    }
  }
}
----
// TEST[setup:my_index]
// TEST[s/my-field/http.response.bytes/]

[discrete]
[[agg-caches]]
=== Aggregation caches

For faster responses, {es} caches the results of frequently run aggregations in
the <<shard-request-cache,shard request cache>>. To get cached results, use the

> same <<shard-and-node-preference,`preference` string>> for each search. {es}

Reviewer: I think that the `preference` is less important than making sure you have `size:0`. If you have two shard copies you'll fill the cache pretty quick without preference. But if you don't have `size:0` you can't get it at all!

Author: I updated this recommending setting `size` to `0` if you don't need search results.

routes searches with the same preference string to the same shards. If the
shards' data doesn’t change between searches, the shards return cached
aggregation results.

[discrete]
[[limits-for-long-values]]
=== Limits for `long` values

When running aggregations, {es} uses <<number,`double`>> values to hold and
represent numeric data. As a result, aggregations on <<number,`long`>> numbers

> greater than +2^53^+ are approximate.

Reviewer: some results are approximate anyway by their very nature. Like cardinality. But I don't have a good way to say that.

Author: > some results are approximate anyway by their very nature

That feels more relevant for the docs of an individual agg. This seems more like a limitation that could apply to any agg dealing with numeric data.

[partintro]
--
An aggregation summarizes your data as metrics, statistics, or other analytics.
Aggregations help you answer questions like:

* What's the average load time for my website?
* Who are my most valuable customers based on transaction volume?
* What would be considered a large file on my network?
* How many products are in each product category?

{es} organizes aggregations into three categories:

* <<search-aggregations-metrics,Metric>> aggregations that calculate metrics,
such as a sum or average, from field values.

* <<search-aggregations-bucket,Bucket>> aggregations that
group documents into buckets, also called bins, based on field values, ranges,
or other criteria.

* <<search-aggregations-pipeline,Pipeline>> aggregations that take input from
other aggregations instead of documents or fields.

[discrete]
[[run-an-agg]]
=== Run an aggregation

> {es} runs aggregations as part of a <<search-your-data,search>>. To run an
> aggregation, use the <<search-search,search API>>'s `aggs` parameter. The

Reviewer 2: ```suggestion
You can run aggregations as part of a <<search-your-data,search>> by specifying the <<search-search,search API>>'s `aggs` parameter. The
```

> following search runs `my-agg-name`, a
> <<search-aggregations-bucket-terms-aggregation,terms aggregation>> on
> `my-field`:

Reviewer 2: The agg name is pretty self-explanatory--I don't think it needs to be called out.
```suggestion
following search runs  a
<<search-aggregations-bucket-terms-aggregation,terms aggregation>> on
`my-field`:
```

[source,console]
----
GET /my-index-000001/_search
{
  "aggs": {
    "my-agg-name": {
      "terms": {
        "field": "my-field"
      }
    }
  }
}
----
// TEST[setup:my_index]
// TEST[s/my-field/http.request.method/]

> Aggregation results are in the response's `aggregations` property:

Reviewer 2: Property or object?
